Add Github Testing Workflow + change snapshot serialization

michavie opened the pull request.

andreibancioiu commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,37 @@
+        uses: actions/checkout@v2
+
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
```
I see it's a commonly used one :+1:

andreibancioiu commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,37 @@
+name: Tests
+
+on: ['push', 'pull_request']
```
Maybe only on pull requests, plus workflow dispatch? So that pushes to arbitrary branches (with no associated PR) do not trigger the event?

E.g. https://github.com/multiversx/mx-sdk-js-core/blob/main/.github/workflows/build.yml#L7

michavie replied:
updated 👍

andreibancioiu reviewed (dismissed):
:rocket:

razvanphp commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,40 @@
+          echo "::add-matcher::${{ runner.tool_cache }}/phpunit.json"
+
+      - name: Install PHP dependencies
+        run: composer update --${{ matrix.dependency-version }} --no-interaction --no-progress --ansi
```
I would add before this the composer caching with `--prefer-dist`:
```
      - name: Get composer cache directory
        id: composer-cache
        run: echo "dir=$(composer config cache-files-dir)" >> $GITHUB_OUTPUT

      - name: Cache dependencies
        uses: actions/cache@v3
        with:
            path: ${{ steps.composer-cache.outputs.dir }}
            key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.lock') }}
            restore-keys: ${{ runner.os }}-composer-
```

michavie replied:
Good point; I would however add optimizations like this at a later point. Tests run currently in ~40s which should be a good start. It also would need to be considered that we don't have `composer.lock` checked into version control

razvanphp replied:
I know, this is not about `composer.lock`, but about the time/speed to download the packages from the internet vs having them locally as cache. Given they are downloaded from github itself, this is indeed very fast (2s).

razvanphp commented:
LGTM, but the tests are currently failing...

michavie committed:
add gh testing workflow

michavie committed:
update testing snapshot method

michavie commented:
@razvanphp The test started to fail after I changed the trigger conditions as @andreibancioiu proposed – for some weird reason.

The reason was the way Carbon instances get serialized in GH Actions env where certain extensions are present, so I changed the serialization method for snapshots to json.

Should all be good now.

razvanphp commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,40 @@
+        os: [ubuntu-latest]
+        php: ['8.1', '8.2', '8.3']
+        dependency-version: [prefer-stable]
+        parallel: ['', '--parallel']
```
Where is this used?

michavie replied:
For Pest's parallel execution: https://pestphp.com/docs/optimizing-tests#content-parallel-testing

razvanphp replied:
but you are not passing this variable to the `composer test` command, shouldn't you use `${{ matrix.parallel }}`?
https://github.com/multiversx/mx-sdk-php/actions/runs/8033201667/job/21943439564#step:6:9

michavie replied:
Good catch @razvanphp, the test output now shows `Parallel` info correctly as a confirmation 👍

michavie committed:
fix workflow test parallelization
